internal/kafka/internal/services/quota/quota_management_list_service.go

```diff
@@ -12,39 +15,58 @@ import (
 }
 
-// don't validate billing accounts when using the quota list
+var _ services.QuotaService = &QuotaManagementListService{}
+
+// ValidateBillingAccount - don't validate billing accounts when using the quota list
 func (q QuotaManagementListService) ValidateBillingAccount(organisationId string, instanceType types.KafkaInstanceType, billingCloudAccountId string, marketplace *string) *errors.ServiceError {
 	return nil
 }
 
-func (q QuotaManagementListService) CheckIfQuotaIsDefinedForInstanceType(username string, organisationId string, instanceType types.KafkaInstanceType) (bool, *errors.ServiceError) {
+// CheckIfQuotaIsDefinedForInstanceType - returns if there is any quota configuration for the given instanceType/billingAccount pair (either organization or service account)
+func (q QuotaManagementListService) CheckIfQuotaIsDefinedForInstanceType(username string, organisationId string, instanceType types.KafkaInstanceType, kafkaBillingModel config.KafkaBillingModel) (bool, *errors.ServiceError) {
 	orgId := organisationId
+	var account quota_management.Account
 	org, orgFound := q.quotaManagementList.QuotaList.Organisations.GetById(orgId)
 	userIsRegistered := false
+	serviceAccountIsRegistered := false
+
 	if orgFound && org.IsUserRegistered(username) {
 		userIsRegistered = true
 	} else {
-		_, userFound := q.quotaManagementList.QuotaList.ServiceAccounts.GetByUsername(username)
-		userIsRegistered = userFound
+		account, serviceAccountIsRegistered = q.quotaManagementList.QuotaList.ServiceAccounts.GetByUsername(username)
 	}
 
-	// allow user defined in quota list to create standard instances
-	if userIsRegistered && instanceType == types.STANDARD {
+	// if the user is registered, check that he has quota defined for the desired instance type
+	if userIsRegistered && org.HasQuotaConfigurationFor(instanceType.String(), kafkaBillingModel.ID) {
 		return true, nil
-	} else if !userIsRegistered && instanceType == types.DEVELOPER { // allow user who are not in quota list to create developer instances
+	}
+
+	// if the serviceAccount is registered, check that he has quota defined for the desired instance type
+	if serviceAccountIsRegistered && account.HasQuotaConfigurationFor(instanceType.String(), kafkaBillingModel.ID) {
+		return true, nil
+	}
+
+	// if the user is not listed, he can create only DEVELOPER instances
+	if !userIsRegistered && !serviceAccountIsRegistered && instanceType.String() == types.DEVELOPER.String() { // allow user who are not in quota list to create developer instances
 		return true, nil
 	}
 
 	return false, nil
 }
 
-func (q QuotaManagementListService) ReserveQuota(kafka *dbapi.KafkaRequest, instanceType types.KafkaInstanceType) (string, *errors.ServiceError) {
+// ReserveQuota - tries to reserve the quota for the received kafka request
+func (q QuotaManagementListService) ReserveQuota(kafka *dbapi.KafkaRequest) (string, *errors.ServiceError) {
 	if !q.quotaManagementList.EnableInstanceLimitControl {
```

Comment:
Note that this flag controls whether the `desired_billing_model` is set in the KafkaRequest object in the database if the `billing_model`  is not specified in the Kafka request payload as its set further down this function.

A bit of an edge case if KFM was ran initially with this disabled and then re-ran it with this enabled. It could affect the calculations below. This probably is unlikely to happen though.